### PR TITLE
Fix Megatron-FSDP optimizer CPU offload and checkpointing

### PR DESCRIPTION
**Pull request by wplf**
## Summary
- Handle Megatron-FSDP DTensor parameters and gradients by operating on local shards before optimizer CPU offload copies.
- Save `HybridDeviceOptimizer` state for `fsdp_dtensor` checkpoints as deterministic FSDP DTensors, skipping duplicate `master_param` entries and attaching local DCP chunk metadata without metadata collectives.
- Allow SWiGLU/GDN optimizer checkpoint preprocessing to split flat optimizer DTensors using the corresponding model parameter metadata, and restore loaded DTensor optimizer state back to local tensors.

Fixes https://github.com/NVIDIA/Megatron-LM/issues/4910.

## Tests
- `uv run isort megatron/core/optimizer/distrib_optimizer.py`
- `python -m py_compile megatron/core/optimizer/distrib_optimizer.py megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py megatron/core/transformer/fsdp_dtensor_checkpoint.py`
- `PYTHONDONTWRITEBYTECODE=1 PYTHONPATH=. uv run python -m pytest tests/unit_tests/test_optimizer_cpu_offloading.py -q` (`72 passed`)
- Local Qwen3.5-VL proxy repro with `optimizer_cpu_offload=True`, `optimizer_offload_fraction=1`, `overlap_cpu_optimizer_d2h_h2d=True`, `use_precision_aware_optimizer=True`, and `fsdp_dtensor` optimizer checkpointing: checkpoint saves at iterations 10 and 12 completed successfully.

### CODE REVIEW

**Comment by wplf** [comment]
/ok to test [5da7d67](https://github.com/NVIDIA/Megatron-LM/pull/4623/commits/5da7d67fdfd63544368a20fdd9074f5fabc5526d)

File: megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py [hunk 1]
```diff
@@ -4,6 +4,20 @@
...
+
+    HAVE_DTENSOR = True
+except ImportError:
+    HAVE_DTENSOR = False
+    DTensor = None
+
+
+def _to_local_if_dtensor(tensor):
+    if HAVE_DTENSOR and isinstance(tensor, DTensor):
+        return tensor.to_local()
```
Comment (cspades) [inline megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py]: BTW, if we don't need differentiability (such as in this optimizer step), sometimes `dtensor._local_tensor` is more robust since we can be sure we do not receive a copy of the original local Tensor. (Don't ask me why I know this.)

**Review by cspades — COMMENTED** [review]
@wplf Could you share some information on what the bug was? Curious what was the root cause and how it is related to DTensors!

File: megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py (hunk 1 above)
Comment (wplf) [reply megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py]: Hi, @cspades .

Currently, when using `megatron-fsdp` and `--optimizer-cpu-offload`, it will crash. Root cause is hybrid-optimizer does not take mfsdp into consideration. Use Dtensor._local_tensor will resolve this issue.

But I haven't check this fix thoroughly for training convergence. 

I'll check training stability and convert `Dtensor.to_local()` to `Dtensor._local_tensor` as soon as possible.

Comment (wplf) [reply megatron/core/optimizer/cpu_offloading/hybrid_optimizer.py]: Addressed in `eff7bfb06`: `_to_local_if_dtensor()` now returns `dtensor._local_tensor` instead of `dtensor.to_local()` for the hybrid optimizer path. Re-ran `python -m py_compile` on the touched checkpoint/optimizer files and `PYTHONDONTWRITEBYTECODE=1 PYTHONPATH=. uv run python -m pytest tests/unit_tests/test_optimizer_cpu_offloading.py -q` (`72 passed`).

File: megatron/core/optimizer/distrib_optimizer.py [hunk 2]
```diff
@@ -1594,15 +1616,167 @@ def sharded_param_state_fsdp_dtensor(self, is_loading: bool = False):
...
+        # each local shard as replicated data. Its fp32 master params duplicate
+        # Megatron-FSDP's model-state main weights, so they are intentionally not
+        # stored in the optimizer state.
+        if isinstance(self.optimizer, HybridDeviceOptimizer):
+            packed_state = self._pack_hybrid_optimizer_fsdp_state_dict()
+        else:
+            packed_state = {
+                (self._param_name(k) if isinstance(k, torch.Tensor) else k): v
+                for k, v in self.state.items()
+            }
```
Comment (cspades) [inline megatron/core/optimizer/distrib_optimizer.py:1631]: This optimizer isn't just used for FSDP right? We should check that the parameters are FSDP parameters before doing this.

File: megatron/core/optimizer/distrib_optimizer.py [hunk 3]
```diff
@@ -1594,15 +1616,167 @@ def sharded_param_state_fsdp_dtensor(self, is_loading: bool = False):
...
+                )
+                continue
+            packed_state[key] = state
+
+        return packed_state
+
+    def _pack_hybrid_optimizer_fsdp_state(
+        self, param: torch.nn.Parameter, param_name: str, state: dict[str, Any]
+    ) -> dict[str, Any]:
+        """Convert HybridDeviceOptimizer FSDP-local tensor state to DTensors."""
```
Comment (cspades) [inline megatron/core/optimizer/distrib_optimizer.py:1673]: Why not just use [`preprocess_state_dict_for_uneven_dtensor` or `update_uneven_dtensor_chunk_metadata`](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/distributed/fsdp/src/megatron_fsdp/uneven_dtensor.py#L238) instead of re-implementing it here? (Also, we should have all of these types of utilities inside the Megatron-FSDP source directory unless it absolutely needs to be in the training loop / optimizer code!)

File: megatron/core/transformer/fsdp_dtensor_checkpoint.py [hunk 4]
```diff
@@ -304,11 +304,14 @@ def split_swiglu_linear_fc1(data, dist_param, swiglu_shard_axis, is_expert_param
...
-            )
-            local_tensor = data.to_local()
+            if data.shape == global_shape:
+                local_tensor = data.to_local()
+            else:
+                assert data.numel() == dist_param.numel(), (
+                    f"DTensor shape mismatch: data.shape={data.shape} vs "
+                    f"dist_param.shape={global_shape}"
+                )
+                local_tensor = data.to_local().view(-1)
```
Comment (cspades) [inline megatron/core/transformer/fsdp_dtensor_checkpoint.py:314]: These can also be `DTensor._local_tensor` BTW! Plus more instances of this in this same file.

File: megatron/core/optimizer/distrib_optimizer.py [hunk 5]
```diff
@@ -1594,15 +1616,167 @@ def sharded_param_state_fsdp_dtensor(self, is_loading: bool = False):
...
+                        value.data.view(-1),
+                        flat_param,
+                        dist_index=param.megatron_fsdp_dist_index,
+                        is_expert_param=is_expert_param,
+                        run_check=False,
+                        update_uneven_dtensor_chunk_meta=False,
+                    )
+                    self._set_flat_fsdp_dtensor_chunk_metadata(
+                        value, param.megatron_fsdp_slice
+                    )
```
Comment (cspades) [inline megatron/core/optimizer/distrib_optimizer.py:1735]: So TL;DR I assume the `packed_state` is our offloaded state, and we want them to be DTensors as well?

File: megatron/core/optimizer/distrib_optimizer.py [hunk 6]
```diff
@@ -1594,15 +1616,167 @@ def sharded_param_state_fsdp_dtensor(self, is_loading: bool = False):
...
+                        f"Failed to pack HybridDeviceOptimizer state '{key}' for "
+                        f"FSDP parameter '{param_name}' on rank {rank}."
+                    ) from exc
+            packed_state[key] = value
+
+        return packed_state
+
+    @staticmethod
+    def _set_flat_fsdp_dtensor_chunk_metadata(tensor: DTensor, fsdp_slice: slice) -> None:
+        """Attach DCP chunk metadata for a flat FSDP-local shard without collectives."""
```
Comment (cspades) [inline megatron/core/optimizer/distrib_optimizer.py:1748]: We can just re-factor the existing function in `uneven_dtensor.py` to skip the metadata AG update! Less code outside MFSDP the better!

**Review by cspades — COMMENTED** [review]
Some important nits, basically if we can move most of this code into MFSDP source it would be much cleaner and better.

INFO: Could you also add more commentary explaining the lifecycle of the offloaded DTensors here? Just want to make sure if this is the right way to implement this.
